### PR TITLE
WIP: ROX-18652: Secured Cluster Upgrade v2 PoC

### PR DESCRIPTION
**Pull request by Maddosaurus**
# THIS IS A POC - DO NOT REVIEW / MERGE

This PoC researches how we can have an independent version control of the Operator outside of OLM.
Steps to reproduce:
1. Have a central with new running endpoint available (e.g. `quay.io/stackrox-io/main:4.1.x-557-g89da63b5f5`)
2. Generate an init bundle on said central, download the Kubernetes files and apply them
3. In a local cluster, deploy Operator via OLM (see `README.md` in `/operator`)
4. Apply `rhacs-operator-account.yaml` to create necessary account and role
5. Update the central endpoint in `scs-cr.yaml` to your central address from step 1, then apply the CR

See comment for the yamls

### CODE REVIEW

**Comment by Maddosaurus** [comment]
Flip open the file references for the full files.

<details>
<summary>scs-cr.yaml</summary>

```yaml
apiVersion: platform.stackrox.io/v1alpha1
kind: SecuredCluster
metadata:
  name: stackrox-secured-cluster-services
spec:
  clusterName: testing-cluster
  centralEndpoint: 35.246.227.223:443
  scanner:
    scannerComponent: Disabled
  admissionControl:
    resources:
      requests:
        memory: 100Mi
        cpu: 100m
  sensor:
    resources:
      requests:
        memory: 100Mi
        cpu: 100m
  perNode:
    collector:
      resources:
        requests:
          memory: 100Mi
          cpu: 100m
    compliance:
      resources:
        requests:
          memory: 100Mi
          cpu: 100m
    nodeInventory:
      resources:
        requests:
          memory: 100Mi
          cpu: 100m
```
</details>

<details>
<summary>rhacs-operator-account.yaml</summary>

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: rhacs-operator-secured-cluster-services
rules:
  - apiGroups:
      - '*'
    resources:
      - '*'
    verbs:
      - '*'
  - apiGroups:
      - platform.stackrox.io
    resources:
      - centrals
    verbs:
      - create
      - delete
      - get
      - list
      - patch
      - update
      - watch
  - apiGroups:
      - platform.stackrox.io
    resources:
      - centrals/finalizers
    verbs:
      - update
  - apiGroups:
      - platform.stackrox.io
    resources:
      - centrals/status
    verbs:
      - get
      - patch
      - update
  - apiGroups:
      - platform.stackrox.io
    resources:
      - securedclusters
    verbs:
      - create
      - delete
      - get
      - list
      - patch
      - update
      - watch
  - apiGroups:
      - platform.stackrox.io
    resources:
      - securedclusters/finalizers
    verbs:
      - update
  - apiGroups:
      - platform.stackrox.io
    resources:
      - securedclusters/status
    verbs:
      - get
      - patch
      - update
  - apiGroups:
      - authentication.k8s.io
    resources:
      - tokenreviews
    verbs:
      - create
  - apiGroups:
      - authorization.k8s.io
    resources:
      - subjectaccessreviews
    verbs:
      - create
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: rhacs-operator-stackrox-secured-cluster-services
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: rhacs-operator-secured-cluster-services
subjects:
  - kind: ServiceAccount
    name: rhacs-operator-controller-manager-stackrox-secured-cluster-services
    namespace: rhacs-inner-operator
---
apiVersion: v1
imagePullSecrets:
  - name: quay-ips
kind: ServiceAccount
metadata:
  name: rhacs-operator-controller-manager-stackrox-secured-cluster-services
  namespace: rhacs-inner-operator

```

</details>

File: operator/pkg/inneroperator/values/translation/translation.go [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+
+	tversion, err := helmUtil.ValuesForKVPair("operator.version", targetVersion)
+	if err != nil {
+		return nil, err
+	}
+	sugar.Infof("tversion: %v", tversion)
+	// Overwrite target image version of Operator with result from central
+	// CHECKME: As this is preferred left-to-right, is this a bug in the other implementations?
+	//			Technically, we would need to apply CRs first, then imageOverrides, then baseValues, no?
+	values := helmUtil.CoalesceTables(tversion, baseValues)
```
Comment (SimonBaeumer) [inline operator/pkg/inneroperator/values/translation/translation.go:89]: Is it intended that the `valsFromCR` are not returned by the `Translate` function?

Comment (Maddosaurus) [reply operator/pkg/inneroperator/values/translation/translation.go:89]: Yes, that is correct. At this point in time, we are only interested in the `centralEndpoint` of the Secured Cluster CR.
As we are in Outer Operator Mode, we mainly process Central CRs fully.
We listen passively to the creation of Secured Cluster CRs, extract only the central endpoint from it and query that endpoint for a target version of the inner Operator to deploy.

**Comment by Maddosaurus** [comment]
Closed as we need to re-architecture the ScU v2 before taking another round of PoC

**Pull request closed without merging**